File: 0123.买卖股票的最佳时机III/0123-买卖股票的最佳时机III.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if prices==None or len(prices)<=1:
            return 0
        firstBuy=float('inf')
        firstSale=0
        secondBuy=float('inf')
        secondSale=0
        for i in range(len(prices)):
            firstBuy=min(firstBuy,prices[i])
            firstSale=max(firstSale,prices[i]-firstBuy)
            secondBuy=min(secondBuy,prices[i]-firstSale)
            secondSale=max(secondSale,prices[i]-secondBuy)
        return secondSale

File: 0123.买卖股票的最佳时机III/0123-买卖股票的最佳时机III.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices==null || prices.length==0) return 0;
        int buyFirst=prices[0];
        int buySecond=prices[0];
        int profitFirst=0;
        int profitSecond=0;
        for (int i=0;i<prices.length;i++){
            buyFirst=Math.min(buyFirst,prices[i]);
            profitFirst=Math.max(profitFirst,prices[i]-buyFirst);
            buySecond=Math.min(buySecond,prices[i]-profitFirst);
            profitSecond=Math.max(profitSecond,prices[i]-buySecond);
        }
        return profitSecond;
    }
}